samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c

```diff
@@ -149,232 +210,116 @@ static void state_binding(u8_t lightness, u8_t temperature)
-		light_lightness_srv_user_data.linear =
-			(u16_t) (65535 * tmp * tmp);
-
-		light_lightness_srv_user_data.last =
-			light_lightness_srv_user_data.actual;
-
-		gen_level_srv_root_user_data.level =
-			light_lightness_srv_user_data.actual - 32768;
-
-		light_ctl_srv_user_data.lightness =
-			light_lightness_srv_user_data.actual;
+		update_gen_onoff_state();
 		break;
 	case ACTUAL: /* Lightness update as per Light Lightness Actual state */
-		tmp = ((float) light_lightness_srv_user_data.actual / 65535);
-		light_lightness_srv_user_data.linear =
-			(u16_t) (65535 * tmp * tmp);
-
-		light_lightness_srv_user_data.last =
-			light_lightness_srv_user_data.actual;
-
-		gen_level_srv_root_user_data.level =
-			light_lightness_srv_user_data.actual - 32768;
-
-		light_ctl_srv_user_data.lightness =
-			light_lightness_srv_user_data.actual;
+		update_gen_onoff_state();
 		break;
 	case LINEAR: /* Lightness update as per Light Lightness Linear state */
 		tmp16 = (u16_t) 65535 *
 			sqrt(((float) light_lightness_srv_user_data.linear /
-			      65535));
+			65535));
 
-		if (tmp16 > 0 && tmp16 <
-		    light_lightness_srv_user_data.lightness_range_min) {
-			tmp16 =
-			light_lightness_srv_user_data.lightness_range_min;
-		} else if (tmp16 >
-			   light_lightness_srv_user_data.lightness_range_max) {
-			tmp16 =
-			light_lightness_srv_user_data.lightness_range_max;
-		}
+		bin = constrain_lightness_actual_state(tmp16);
+		update_gen_onoff_state();
 
-		light_lightness_srv_user_data.actual = tmp16;
-
-		tmp = ((float) light_lightness_srv_user_data.actual / 65535);
-		light_lightness_srv_user_data.linear =
-			(u16_t) (65535 * tmp * tmp);
-
-		light_lightness_srv_user_data.last =
-			light_lightness_srv_user_data.actual;
-
-		gen_level_srv_root_user_data.level =
-			light_lightness_srv_user_data.actual - 32768;
-
-		light_ctl_srv_user_data.lightness =
-			light_lightness_srv_user_data.actual;
+		if (bin) {
+			goto ignore_linear_update_others;
+		}
 		break;
 	case CTL: /* Lightness update as per Light CTL Lightness state */
-		tmp16 = light_ctl_srv_user_data.lightness;
-
-		if (tmp16 > 0 && tmp16 <
-		    light_lightness_srv_user_data.lightness_range_min) {
-			tmp16 =
-			light_lightness_srv_user_data.lightness_range_min;
-		} else if (tmp16 >
-			   light_lightness_srv_user_data.lightness_range_max) {
-			tmp16 =
-			light_lightness_srv_user_data.lightness_range_max;
-		}
-
-		light_lightness_srv_user_data.actual = tmp16;
-
-		tmp = ((float) light_lightness_srv_user_data.actual / 65535);
-		light_lightness_srv_user_data.linear =
-			(u16_t) (65535 * tmp * tmp);
-
-		light_lightness_srv_user_data.last =
-			light_lightness_srv_user_data.actual;
-
-		gen_level_srv_root_user_data.level =
-			light_lightness_srv_user_data.actual - 32768;
-
-		light_ctl_srv_user_data.lightness =
-			light_lightness_srv_user_data.actual;
+		constrain_lightness_actual_state(
+			light_ctl_srv_user_data.lightness);
```

Comment:
This looks quite messed up as well. I'd suggest rethinking your function and variable naming, since they're super long right now.

Reply:
Going to rename as ......constrain_light_actual_state

Reply:
done !!